tool/ci/pkg/bot/bot.go

```diff
@@ -0,0 +1,157 @@
+		},
+	}, nil
+}
+
+// CheckAndSetDefaults verifies configuration and sets defaults
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Environment == nil {
+		return trace.BadParameter("missing parameter Environment")
+	}
+	return nil
+}
+
+// invalidateApprovals dismisses all approved reviews on a pull request.
+func (c *Bot) invalidateApprovals(ctx context.Context, msg string, reviews []review) error {
+	pr := c.Environment.PullRequest
+	for _, v := range reviews {
+		if v.status == ci.Approved {
+			_, _, err := c.Environment.Client.PullRequests.DismissReview(ctx,
+				pr.RepoOwner,
+				pr.RepoName,
+				pr.Number,
+				v.id,
+				&github.PullRequestReviewDismissalRequest{Message: &msg},
+			)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+	return nil
+}
+
+// DimissStaleWorkflowRunsForExternalContributors dismisses stale workflow runs for external contributors
+func (gc GithubClient) DimissStaleWorkflowRunsForExternalContributors(ctx context.Context, token, repoOwner, repoName string) error {
+	pulls, _, err := gc.Client.PullRequests.List(ctx, repoOwner, repoName, &github.PullRequestListOptions{State: ci.Open})
+	if err != nil {
+		return err
+	}
+	for _, pull := range pulls {
+		err := gc.DismissStaleWorkflowRuns(ctx, token, *pull.Base.User.Login, *pull.Base.Repo.Name, *pull.Head.Ref)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// DismissStaleWorkflowRuns dismisses stale Check workflow runs.
+// Stale workflow runs are workflow runs that were previously ran and are no longer valid
+// due to a new event triggering thus a change in state. The workflow running in the current context is the source of truth for
+// the state of checks.
+func (gc GithubClient) DismissStaleWorkflowRuns(ctx context.Context, token, owner, repoName, branch string) error {
+	var targetWorkflow *github.Workflow
+	workflows, _, err := gc.Client.Actions.ListWorkflows(ctx, owner, repoName, &github.ListOptions{})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	for _, w := range workflows.Workflows {
+		if *w.Name == ci.CheckWorkflow {
+			targetWorkflow = w
+			break
+		}
+	}
+	list, _, err := gc.Client.Actions.ListWorkflowRunsByID(ctx, owner, repoName, *targetWorkflow.ID, &github.ListWorkflowRunsOptions{Branch: branch})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	sort.Sort(ByTime(list.WorkflowRuns))
+
+	// Deleting all runs except the most recent one.
+	for i := 0; i < len(list.WorkflowRuns)-1; i++ {
+		run := list.WorkflowRuns[i]
+		err := gc.deleteRun(ctx, token, owner, repoName, *run.ID)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// deleteRun deletes a workflow run.
+// Note: the go-github client library does not support this endpoint.
+func (gc GithubClient) deleteRun(ctx context.Context, token, owner, repo string, runID int64) error {
+	// Construct url
+	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/actions/runs/%v", owner, repo, runID)
```

Comment:
Construct the URL: https://pkg.go.dev/net/url#URL, I think that will also automatically do path cleanup for you as well.